### PR TITLE
File write

### PR DESCRIPTION
**Pull request by dariajung**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,12 +5,13 @@
   "repository": "https://github.com/angular/ts-minify",
   "license": "Apache-2.0",
   "scripts": {
-    "prepublish": "npm install tsd@^0.6.3 && tsd install node && tsd install mocha && tsd install chai",
+    "prepublish": "npm install tsd@^0.6.3 && tsd install node && tsd install mocha && tsd install chai && tsd install fs-extra",
```
Comment (mprobst) [inline package.json]: have you considered adding a `tsd.json` file to contain all these required typings?

File: src/main.ts [hunk 2]
```diff
@@ -59,9 +63,30 @@ export class Minifier {
...
+          var renamedTSCode = this.visit(f);
+          var fileName = this.getOutputPath(f.fileName, destination);
+          fsx.mkdirsSync(path.dirname(fileName));
+          fs.writeFileSync(fileName, renamedTSCode);
+        });
+  }
+
+  getOutputPath(filePath: string, destination: string): string {
+    var parsedFile = path.parse(filePath);
+    var renamedFile = `${destination}/${parsedFile.name}${parsedFile.ext}`;
```
Comment (mprobst) [inline src/main.ts]: What happens if you have as inputs `first/b.ts` and `second/b.ts`?

File: package.json (hunk 1 above)
Comment (dariajung) [inline package.json]: Added a `tsd.json`

**Comment by mprobst** [comment]
Is this ready for another review?

**Comment by dariajung** [comment]
@mprobst Yes.

File: src/main.ts [hunk 3]
```diff
@@ -22,8 +27,9 @@ export class Minifier {
   private lastGeneratedPropName: string = '';
   private typeChecker: ts.TypeChecker;
   private errors: string[] = [];
+  private minifierOptions: MinifierOptions = {};
 
-  constructor() {}
+  constructor(options: MinifierOptions = {}) { this.minifierOptions = options; }
```
Comment (mprobst) [inline src/main.ts]: you can declare this as:

```ts
constructor(private options: MinifierOptions = {}) {}
```

Then you need neither the field declaration nor the assignment within the ctor.

File: src/main.ts [hunk 4]
```diff
@@ -54,14 +60,48 @@ export class Minifier {
...
+          var renamedTSCode = this.visit(f);
+          var fileName = this.getOutputPath(f.fileName, destination);
+          fsx.mkdirsSync(path.dirname(fileName));
+          fs.writeFileSync(fileName, renamedTSCode);
+        });
+  }
+
+  getOutputPath(filePath: string, destination?: string): string {
+    // convert everything to absolute paths
+    var destination = path.resolve(process.cwd(), destination || '.');
```
Comment (mprobst) [inline src/main.ts]: I don't think you need to make the destination absolute, do you?

File: src/main.ts [hunk 5]
```diff
@@ -54,14 +60,48 @@ export class Minifier {
...
+    var filePath = path.resolve(process.cwd(), filePath);
+
+    // no base path, flatten file structure and output to destination
+    if (!this.minifierOptions.basePath) {
+      return path.join(destination, path.basename(filePath));
+    }
+
+    if (!path.isAbsolute(this.minifierOptions.basePath)) {
```
Comment (mprobst) [inline src/main.ts]: Not needed either, afaict.

File: src/main.ts [hunk 6]
```diff
@@ -54,14 +60,48 @@ export class Minifier {
...
+      this.minifierOptions.basePath = path.resolve(process.cwd(), this.minifierOptions.basePath);
+    }
+
+    // given a base path, preserve file directory structure
+    var subFilePath = filePath.replace(this.minifierOptions.basePath, '');
```
Comment (mprobst) [inline src/main.ts]: what if the basePath occurs somewhere within the filePath, but not at the beginning?

File: src/main.ts (hunk 3 above)
Comment (dariajung) [inline src/main.ts]: Fixed

**Comment by dariajung** [comment]
@mprobst PTAL

File: src/main.ts [hunk 7]
```diff
@@ -54,14 +59,45 @@ export class Minifier {
...
+          var renamedTSCode = this.visit(f);
+          var fileName = this.getOutputPath(f.fileName, destination);
+          fsx.mkdirsSync(path.dirname(fileName));
+          fs.writeFileSync(fileName, renamedTSCode);
+        });
+  }
+
+  getOutputPath(filePath: string, destination?: string): string {
+    // convert everything to absolute paths
+    var destination = destination || '.';
```
Comment (mprobst) [inline src/main.ts]: nit: you could make this the default value of the parameter above.

**Comment by mprobst** [comment]
LGTM.

File: src/main.ts (hunk 7 above)
Comment (dariajung) [inline src/main.ts]: Ah, changed to `destination: string = '.'`

**Pull request closed without merging**